sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeServiceClient.cs

```diff
@@ -0,0 +1,568 @@
+        /// </param>
+        /// <param name="credential">
+        /// The token credential used to sign requests.
+        /// </param>
+        /// <param name="options">
+        /// Optional client options that define the transport pipeline
+        /// policies for authentication, retries, etc., that are applied to
+        /// every request.
+        /// </param>
+        public DataLakeServiceClient(Uri serviceUri, TokenCredential credential, DataLakeClientOptions options = default)
+            : this(serviceUri, credential.AsPolicy(), options ?? new DataLakeClientOptions())
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DataLakeServiceClient"/>
+        /// class.
+        /// </summary>
+        /// <param name="serviceUri">
+        /// A <see cref="Uri"/> referencing the Data Lake service
+        /// </param>
+        /// <param name="authentication">
+        /// An optional authentication policy used to sign requests.
+        /// </param>
+        /// <param name="options">
+        /// Optional client options that define the transport pipeline
+        /// policies for authentication, retries, etc., that are applied to
+        /// every request.
+        /// </param>
+        internal DataLakeServiceClient(Uri serviceUri, HttpPipelinePolicy authentication, DataLakeClientOptions options)
+            : this(serviceUri, authentication, options, new ClientDiagnostics(options))
+        {
+
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DataLakeServiceClient"/>
+        /// class.
+        /// </summary>
+        /// <param name="serviceUri">
+        /// A <see cref="Uri"/> referencing the Data Lake service.
+        /// </param>
+        /// <param name="authentication">
+        /// An optional authentication policy used to sign requests.
+        /// </param>
+        /// <param name="options">
+        /// Optional client options that define the transport pipeline
+        /// policies for authentication, retries, etc., that are applied to
+        /// every request.
+        /// </param>
+        /// <param name="clientDiagnostics"></param>
+        internal DataLakeServiceClient(Uri serviceUri, HttpPipelinePolicy authentication, DataLakeClientOptions options, ClientDiagnostics clientDiagnostics)
+        {
+
+            _pipeline = (options ?? new DataLakeClientOptions()).Build(authentication);
+            _uri = serviceUri;
+            _blobUri = GetBlobUri(serviceUri);
+            _blobServiceClient = new BlobServiceClient(_blobUri, authentication, options);
+            _clientDiagnostics = clientDiagnostics;
+        }
+        #endregion ctors
+
+        /// <summary>
+        /// Gets the blob Uri.
+        /// </summary>
+        private static Uri GetBlobUri(Uri uri)
+        {
+            Uri blobUri;
+            if (uri.Host.Contains(Constants.DataLake.DfsUriSuffex))
```

Comment:
This doesn't feel sufficient because `Dfs` in my account name would trigger it.  Could we write a `DataLakeUriBuilder` that takes this apart a little safer?

Reply:
I think this is a good idea.

The overall problem I'm trying to solved with the Uris is that we need the DFS Uri, the Blob Uri, and want to return the original Uri the customer provided when they call DataLakeServiceClient.Uri.

Do you think we should have a DataLakeUriBuilder and a BlobUriBuilder in the DataLake namespace?

Reply:
https://github.com/Azure/azure-sdk-for-net/issues/8386

Reply:
I'd just add a DataLakeUriBuilder and have ToDataLakeUri and ToBlobUri methods on it that puts the prefix in the right location.